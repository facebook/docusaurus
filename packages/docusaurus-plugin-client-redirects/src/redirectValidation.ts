/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Joi, PathnameSchema} from '@docusaurus/utils-validation';
import type {RedirectItem} from './types';

const RedirectSchema = Joi.object<RedirectItem>({
  from: PathnameSchema.required(),
  to: Joi.string().required(),
});

export function validateRedirect(redirect: RedirectItem): void {
  const {error} = RedirectSchema.validate(redirect, {
    abortEarly: true,
    convert: false,
  });

  if (error) {
    // Tells the user which redirect is the problem!
    throw new Error(
      `${JSON.stringify(redirect)} => Validation error: ${error.message}`,
    );
  }
}
