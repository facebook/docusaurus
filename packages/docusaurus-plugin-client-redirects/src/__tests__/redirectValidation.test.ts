/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {validateRedirect} from '../redirectValidation';

describe('validateRedirect', () => {
  it('validate good redirects without throwing', () => {
    expect(() => {
      validateRedirect({
        from: '/fromSomePath',
        to: '/toSomePath',
      });
      validateRedirect({
        from: '/from/Some/Path',
        to: '/toSomePath',
      });
      validateRedirect({
        from: '/fromSomePath',
        to: '/toSomePath',
      });
      validateRedirect({
        from: '/fromSomePath',
        to: '/to/Some/Path',
      });
      validateRedirect({
        from: '/fromSomePath',
        to: '/toSomePath?a=1',
      });
      validateRedirect({
        from: '/fromSomePath',
        to: '/toSomePath#anchor',
      });
      validateRedirect({
        from: '/fromSomePath',
        to: '/toSomePath?a=1&b=2#anchor',
      });
    }).not.toThrow();
  });

  it('throw for bad redirects', () => {
    expect(() =>
      validateRedirect({
        from: 'https://fb.com/fromSomePath',
        to: '/toSomePath',
      }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      validateRedirect({
        from: '/fromSomePath?a=1',
        to: '/toSomePath',
      }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      validateRedirect({
        from: '/fromSomePath#anchor',
        to: '/toSomePath',
      }),
    ).toThrowErrorMatchingSnapshot();
  });
});
