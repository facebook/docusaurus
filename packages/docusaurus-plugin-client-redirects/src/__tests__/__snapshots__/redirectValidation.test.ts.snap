// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRedirect throw for bad redirects 1`] = `"{"from":"https://fb.com/fromSomePath","to":"/toSomePath"} => Validation error: "from" (https://fb.com/fromSomePath) is not a valid pathname. Pathname should start with slash and not contain any domain or query string."`;

exports[`validateRedirect throw for bad redirects 2`] = `"{"from":"/fromSomePath?a=1","to":"/toSomePath"} => Validation error: "from" (/fromSomePath?a=1) is not a valid pathname. Pathname should start with slash and not contain any domain or query string."`;

exports[`validateRedirect throw for bad redirects 3`] = `"{"from":"/fromSomePath#anchor","to":"/toSomePath"} => Validation error: "from" (/fromSomePath#anchor) is not a valid pathname. Pathname should start with slash and not contain any domain or query string."`;
